agent/run_job.go

```diff
@@ -0,0 +1,233 @@
+	cctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	defer r.cleanup(cctx, &wg, exit)
+
+	// Before executing the bootstrap process with the received Job env, execute the pre-bootstrap hook (if present) for
+	// it to tell us whether it is happy to proceed.
+	if hook, _ := hook.Find(r.conf.AgentConfiguration.HooksPath, "pre-bootstrap"); hook != "" {
+		// Once we have a hook any failure to run it MUST be fatal to the job to guarantee a true positive result from the hook
+		ok, err := r.executePreBootstrapHook(ctx, hook)
+		if !ok {
+			// Ensure the Job UI knows why this job resulted in failure
+			r.logStreamer.Process([]byte("pre-bootstrap hook rejected this job, see the buildkite-agent logs for more details"))
+			// But disclose more information in the agent logs
+			r.logger.Error("pre-bootstrap hook rejected this job: %s", err)
+
+			exit.Status = "-1"
+			exit.SignalReason = "agent_refused"
+
+			return nil
+		}
+	}
+
+	// Kick off log streaming and job status checking when the process starts.
+	wg.Add(2)
+	go r.jobLogStreamer(cctx, &wg)
+	go r.jobCancellationChecker(cctx, &wg)
+
+	exit = r.runJob(cctx) // Ignore gostaticcheck here, the value of exit is captured by the deferred cleanup function above
+
+	return nil
+}
+
+type processExit struct {
+	Status       string
+	Signal       string
+	SignalReason string
+}
+
+func (r *JobRunner) runJob(ctx context.Context) *processExit {
+	exit := &processExit{}
+	// Run the process. This will block until it finishes.
+	if err := r.process.Run(ctx); err != nil {
+		// Send the error as output
+		r.logStreamer.Process([]byte(err.Error()))
+
+		// The process did not run at all, so make sure it fails
+		return &processExit{
+			Status:       "-1",
+			SignalReason: "process_run_error",
+		}
+	}
+	// Intended to capture situations where the job-exec (aka bootstrap) container did not
+	// start. Normally such errors are hidden in the Kubernetes events. Let's feed them up
+	// to the user as they may be the caused by errors in the pipeline definition.
+	k8sProcess, ok := r.process.(*kubernetes.Runner)
+	if ok && r.cancelled && !r.stopped && k8sProcess.ClientStateUnknown() {
+		r.logStreamer.Process([]byte(
+			"Some containers had unknown exit statuses. Perhaps they were in ImagePullBackOff.",
+		))
+	}
+
+	// Add the final output to the streamer
+	r.logStreamer.Process(r.output.ReadAndTruncate())
+
+	// Collect the finished process' exit status
+	exit.Status = fmt.Sprintf("%d", r.process.WaitStatus().ExitStatus())
+
+	if ws := r.process.WaitStatus(); ws.Signaled() {
+		exit.Signal = process.SignalString(ws.Signal())
+	}
+
+	if r.stopped {
+		// The agent is being gracefully stopped, and we signaled the job to end. Often due
+		// to pending host shutdown or EC2 spot instance termination
+		exit.SignalReason = "agent_stop"
+	} else if r.cancelled {
+		// The job was signaled because it was cancelled via the buildkite web UI
+		exit.SignalReason = "cancel"
+	}
```

Comment:
The urge to turn this into a switch is strong, but I must resist...

...Wait, why must I resist?

```suggestion
	switch {
	case r.stopped:
		// The agent is being gracefully stopped, and we signaled the job to end. Often due
		// to pending host shutdown or EC2 spot instance termination
		exit.SignalReason = "agent_stop"

	case r.cancelled:
		// The job was signaled because it was cancelled via the buildkite web UI
		exit.SignalReason = "cancel"
	}
```